[inductor] Refactor device guard Python codegen to allow nested indentation

aakhundov opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119759
* __->__ #119673

Summary: The codegen of `with torch.cuda._DeviceGuard` context manager in the Python wrapper code is implemented via `device_cm_stack: contextlib.ExitStack()`. As the context managers in the stack are `code.indent()`, this means that the whole stack is unindented at once on `device_cm_stack.close()`. This becomes problematic when attempting to codegen indented code (e.g., for control flow in Python and / or nested subgraph codegen-ing).

In this PR, we refactor the device guard codegen-ing in Python by replacing the `device_cm_stack` by explicit indent and unindent calls for entering and exiting the `with torch.cuda._DeviceGuard` context manager. This allows for nested device guard context managers and better aligns with other indented codegen-ing intertwined with it (e.g., for nested subgraph codegen-ing).

This is necessary for the upcoming support for `torch.cond` (and other control flow operators) in Inductor. Before that, the only change in the Python wrapper codegen is that the `return outputs` is now happening outside the `with torch.cuda._DeviceGuard` context manager.

Test Plan: CI

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler

aakhundov committed:
[inductor] Remove with torch.cuda._DeviceGuard from wrapper codegen

Summary: As we are already codegen-ing `set_device` on CUDA device change, `with torch.cuda._DeviceGuard` seems redundant. Its "multiple enters, one exit" indentation structure conflicts with (arbitrarily nested) indented control flow codegen in Python. To this end, we try removing `with torch.cuda._DeviceGuard` in this PR.

Test Plan: CI

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -282,19 +282,14 @@ def codegen(
-            # integers
-            code.writeline(f"with {V.graph.device_ops.device_guard(self.device_idx)}:")
-            device_cm_stack.enter_context(code.indent())
             code.writeline(V.graph.device_ops.set_device(self.device_idx))
```
Don't we need to restore the original active device when exiting the wrapper code? Both for normal exits and on exceptions.

aakhundov replied:
Good point. Let me look further into this.

aakhundov committed:
Update on "[inductor] Remove with torch.cuda._DeviceGuard from wrapper codegen"


Summary: As we are already codegen-ing `set_device` on CUDA device change, `with torch.cuda._DeviceGuard` seems redundant. Its "multiple enters, one exit" indentation structure conflicts with (arbitrarily nested) indented control flow codegen in Python. To this end, we try removing `with torch.cuda._DeviceGuard` in this PR.

Test Plan: CI

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler

[ghstack-poisoned]

aakhundov commented:
@pytorchbot merge
